### PR TITLE
Github action build and unit test workflows

### PR DESCRIPTION
**Pull request by cea2aj**
Create callable build and unit test workflows for GH actions

These are split into callable workflows so that they can be reused for other workflows (e.g. i18n builds, and search bar assets)

Note: A current limitation of callable workflows is that they currently need to specify the full repo and the branch. The PM for Github Actions said that they would be releasing an update in a couple of weeks which would allow us to call workflows from within the same repo. We can keep this migration in a feature branch until they release that feature and then we can update the code and complete the migration

Some additional optimizations include adding caching to NPM node modules and setting up concurrency so that new commits will cancel the GH action jobs for previous commits

J=SLAP-1811, SLAP-1812
TEST=manual

Observe that in the 'Actions' tab the build and unit tests run successfully

### CODE REVIEW

File: .github/workflows/build_and_deploy.yml
Comment (oshi97) [inline .github/workflows/build_and_deploy.yml:15]: is it okay that this is always pointing to `dev/build-workflow`?

File: .github/workflows/build.yml
Comment (oshi97) [inline .github/workflows/build.yml:4]: this overall process seems cool!

File: .github/workflows/build_and_deploy.yml
Comment (cea2aj) [reply .github/workflows/build_and_deploy.yml:15]: No, but the PM for Github Actions reusable workflows said that it will soon be possible to have it point to the same branch: https://github.community/t/ref-head-in-reusable-workflows/203690 Since we aren't in a rush to complete this migration, I was thinking we could hold off on completing the migration until the feature is released. We can point the workflows at the feature branch in the mean time (I can update to point to the feature branch once these workflows are merged into the feature branch)

File: .github/workflows/build_and_deploy.yml
Comment (yen-tt) [inline .github/workflows/build_and_deploy.yml:7]: how come we are ignoring master for this build and deploy workflow?

Comment (cea2aj) [reply .github/workflows/build_and_deploy.yml:7]: This is based on our circle CI config where master has a separate deploy process (build_and_hold) which requires manual approval. That config is scheduled for future items

Comment (yen-tt) [reply .github/workflows/build_and_deploy.yml:7]: ah gotcha

File: .github/workflows/build_and_deploy.yml
Comment (yen-tt) [inline .github/workflows/build_and_deploy.yml:22]: what's the `-1` for?

Comment (oshi97) [reply .github/workflows/build_and_deploy.yml:22]: I think that's something like a version number, similar to our `v3` in our circleci cache key

Comment (cea2aj) [reply .github/workflows/build_and_deploy.yml:22]: It's just a version number that we can increment if we ever need to make a new group key for the same github ref. It's not really necessary but it also doesn't hurt to have it. It's similar to a cache key, and it's common to have a number on cache keys so that you can manually bust the cache.

**Review by yen-tt — APPROVED** [review]
apart from waiting for the feature release to update workflow path to use local reference from current branch instead of dev/build-workflow. lgtm!

**Comment by cea2aj** [comment]
> apart from waiting for the feature release to update workflow path to use local reference from current branch instead of dev/build-workflow. lgtm!

yep, i'll keep the work in the feature branch for now and update in the next PR to point to the feature branch

**Pull request merged**
